### PR TITLE
[5G: MVC][component: agw][type: enhancement] [AMF] Set interface client for AMF towards SMF

### PR DESCRIPTION
**Pull request by ashish-acl**
Signed-off-by: ashish-acl <ashish.t@altencalsoftlabs.com>

## Summary
Added gRPC set interface client implementation for AMF to invoke SetAmfSessionContext procedure at SMF.
   

## Test Plan
[amf_gRPC_client_UT.log](https://github.com/magma/magma/files/5390894/amf_gRPC_client_UT.log)
[PR_3272_make_test_results.log](https://github.com/magma/magma/files/5656961/PR_3272_make_test_results.log)
[PR_3272_LTE_integration_test_results.log](https://github.com/magma/magma/files/5656963/PR_3272_LTE_integration_test_results.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "SmfServiceClient.h"
+#include "ServiceRegistrySingleton.h"
+#include <google/protobuf/util/time_util.h>
+
+using grpc::Status;
+
+namespace {
+
+void callback_void_smf(
```
Comment (sureshgorijavolu) [inline lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp]: can the name be little more meaningful?

Comment (ashish-acl) [reply lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp]: changed the name to SetAmfSessionContextRpcCallback()

File: lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+AsyncSmfServiceClient::AsyncSmfServiceClient()
+    : AsyncSmfServiceClient(
+          magma::ServiceRegistrySingleton::Instance()->GetGrpcChannel(
+              "sessiond", magma::ServiceRegistrySingleton::LOCAL)) {}
+
+bool AsyncSmfServiceClient::set_smf_session(
+    const SetSMSessionContext& request) {
+  set_smf_session_rpc(request, SetAmfSessionContextRpcCallback);
+  return true;
```
Comment (ulaskozat) [inline lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp:42]: Please handle negative return case from the rpc call. Put an error log message.

File: lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#include "ServiceRegistrySingleton.h"
+#include <google/protobuf/util/time_util.h>
+
+using grpc::Status;
+
+namespace {
+
+void SetAmfSessionContextRpcCallback(
+    grpc::Status status, magma::lte::SmContextVoid response) {}
+
```
Comment (ulaskozat) [inline lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp:30]: Please handle negative return case from the rpc call. Put an error log message.

Comment (ulaskozat) [reply lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp:30]: e.g., check the status, if not okay then print out the error in error log. check the response message itself if it has any error codes, etc.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
Please handle the cases with proper error logging for failed grpc calls. also include `\\ TODO` in proper places.

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp (hunk 3 above)
Comment (ashish-acl) [reply lte/gateway/c/oai/lib/n11/SmfServiceClient.cpp:30]: Addressed

**Pull request merged**